lib/gc/Transforms/Microkernel/ConvertLinalgToMicrokernel.cpp

```diff
@@ -0,0 +1,390 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/IR/AffineExprVisitor.h"
+#include "mlir/IR/AffineMap.h"
+#include "llvm/ADT/SetOperations.h"
+#include "llvm/ADT/SmallBitVector.h"
+#include "llvm/ADT/SmallVector.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Rewrite/FrozenRewritePatternSet.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "gc/Dialect/Linalgx/LinalgxOps.h"
+#include "gc/Transforms/Microkernel/MicrokernelPasses.h"
+#include "gc/Transforms/Utils/StructuredOpMatcher.h"
+#include "gc/Transforms/Utils/ValueUtils.h"
+
+namespace mlir::microkernel {
+#define GEN_PASS_DEF_CONVERTLINALGTOMICROKERNEL
+#include "gc/Transforms/Microkernel/MicrokernelPasses.h.inc"
+
+#define DEBUG_TYPE "convert-linalg-to-microkernel"
+
+struct BrgemmInfo {
+  enum BrgemmMode { STRIDE_MODE, LIST_MODE };
+  int64_t m;
+  int64_t n;
+  int64_t k;
+  int64_t batchSize;
+  int64_t addrLen;
+
+  int64_t lda;
+  int64_t ldb;
+  int64_t ldc;
+  int64_t strideA;
+  int64_t strideB;
+
+  bool isInitOutput;
+  BrgemmMode mode;
+};
+
+FailureOr<linalg::ContractionDimensions>
+customInferContractionDims(linalg::LinalgOp linalgOp) {
+  auto dims = linalg::inferContractionDims(linalgOp);
+  if (failed(dims))
+    return dims;
+  if (llvm::isa<linalgx::BatchReduceMatmulVnniOp>(linalgOp)) {
+    // For VnniOp, the K reduction dims (dim index 3 & 4) cannot be infered by
+    // linalg utils because they form complex affine in operand A; Manually add
+    // them here
+    dims->k.push_back(3);
+    dims->k.push_back(4);
+  }
+  return dims;
+}
+
+static bool isMatchingAffineResult(linalg::LinalgOp linalgOp, AffineExpr expr,
+                                   ArrayRef<unsigned> dimPos) {
+  if (dimPos.size() > 2) {
+    return false;
+  }
+  auto firstDim = getAffineDimExpr(dimPos[0], linalgOp.getContext());
+  if (dimPos.size() == 1)
+    return firstDim == expr;
+
+  // If not regular dim affine, check for VNNI format K affine
+  auto secondKPosDim = getAffineDimExpr(dimPos[1], linalgOp.getContext());
+  // An K affine result for VNNI should be this format:
+  // d{kPos[0]} * s{kPos[1]} + d{kPos[1]} (k0 * K_vnni + k1)
+  if (auto add = dyn_cast<AffineBinaryOpExpr>(expr)) {
+    if (add.getKind() == AffineExprKind::Add) {
+      auto lhs = add.getLHS();
+      auto rhs = add.getRHS();
+      if (rhs == secondKPosDim) {
+        auto mul = dyn_cast<AffineBinaryOpExpr>(lhs);
+        if (mul && mul.getKind() == AffineExprKind::Mul &&
+            mul.getLHS() == firstDim) {
+          if (auto cst_affine = dyn_cast<AffineConstantExpr>(mul.getRHS())) {
+            if (cst_affine.getValue() == 2 || cst_affine.getValue() == 4) {
+              return true;
+            }
+          }
+        }
+      }
+    }
+  }
```

Comment:
Please follow https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

Reply:
Fixed.